skbio/io/format/binary_dm.py

```diff
@@ -0,0 +1,178 @@
+|           |                     |Binary DM format              |
++-----------+---------------------+------------------------------+
+|matrix     |float632 or float64  |A (N, N) dataset containing   |
+|           |                     |the values of the             |
+|           |                     |dissimilarity matrix          |
++-----------+---------------------+------------------------------+
+|order      |string               |A (N,) dataset of the sample  |
+|           |                     |IDs, where N is the total     |
+|           |                     |number of IDs                 |
++-----------+---------------------+------------------------------+
+
+.. note:: This file format is most useful for storing large matrices that do
+   not need to be represented in a human-readable format. This format is
+   especially appropriate for facilitating random access to entries in the
+   distance matrix, such as when calculating within and between distances for a
+   subset of samples in a large matrix.
+
+References
+----------
+.. [1] http://www.hdfgroup.org/
+"""
+
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import h5py
+
+from skbio.io import create_format
+from skbio.stats.distance import DissimilarityMatrix, DistanceMatrix
+
+
+binary_dm = create_format('binary_dm', encoding='binary')
+_vlen_dtype = h5py.special_dtype(vlen=str)
+
+
+@binary_dm.sniffer()
+def _binary_dm_sniffer(fh):
+    try:
+        f = h5py.File(fh, 'r')
+    except OSError:
+        return False, {}
+
+    header = _get_header(f)
+    if header is None:
+        return False, {}
+
+    ids = f.get('order')
+    if ids is None:
+        return False, {}
+
+    mat = f.get('matrix')
+    if mat is None:
+        return False, {}
+
+    n = len(ids)
+    if mat.shape != (n, n):
+        return False, {}
+
+    return True, {}
+
+
+@binary_dm.reader(DissimilarityMatrix)
+def _binary_dm_to_dissimilarity(fh):
+    return _h5py_mat_to_skbio_mat(fh)
+
+
+@binary_dm.reader(DistanceMatrix)
+def _binary_dm_to_distance(fh):
+    return _h5py_mat_to_skbio_mat(fh)
+
+
+@binary_dm.writer(DissimilarityMatrix)
+def _dissimilarity_to_binary_dm(obj, fh):
+    return _skbio_mat_to_h5py_mat(fh)
+
+
+@binary_dm.writer(DistanceMatrix)
+def _distance_to_binary_dm(obj, fh):
+    return _skbio_mat_to_h5py_mat(fh)
+
+
+def _h5py_mat_to_skbio_mat(cls, fh):
+    return cls(fh['matrix'], _parse_ids(fh['order']))
+
+
+def _skbio_mat_to_h5py_mat(obj, fh):
+    _set_header(fh)
+
+    ids = fh.create_dataset('order', shape=(len(obj.ids), ), dtype=_vlen_dtype)
+    ids[:] = obj.ids
+    fh.create_dataset('matrix', data=obj.data)
```

Comment:
Here it looks like the usage of `_skbio_mat_to_h5py_mat(fh)` in the `binary_dm` writers does not match the signature `_skbio_mat_to_h5py_mat(obj, fh)`

Reply:
ah, good catch!